validator/client/src/main/java/tech/pegasys/teku/validator/client/restapi/apis/SetGasLimit.java

```diff
@@ -0,0 +1,126 @@
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import java.util.Objects;
+import java.util.Optional;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.infrastructure.json.types.CoreTypes;
+import tech.pegasys.teku.infrastructure.json.types.DeserializableTypeDefinition;
+import tech.pegasys.teku.infrastructure.restapi.endpoints.EndpointMetadata;
+import tech.pegasys.teku.infrastructure.restapi.endpoints.RestApiEndpoint;
+import tech.pegasys.teku.infrastructure.restapi.endpoints.RestApiRequest;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.validator.client.BeaconProposerPreparer;
+import tech.pegasys.teku.validator.client.SetGasLimitException;
+
+public class SetGasLimit extends RestApiEndpoint {
+
+  public static final String ROUTE = "/eth/v1/validator/{pubkey}/gas_limit";
+  private final Optional<BeaconProposerPreparer> beaconProposerPreparer;
+
+  private static final DeserializableTypeDefinition<SetGasLimit.SetGasLimitBody>
+      GAS_LIMIT_REQUEST_BODY =
+          DeserializableTypeDefinition.object(SetGasLimit.SetGasLimitBody.class)
+              .name("SetGasLimitBody")
+              .initializer(SetGasLimit.SetGasLimitBody::new)
+              .withField(
+                  "gas_limit",
+                  CoreTypes.UINT64_TYPE,
+                  SetGasLimit.SetGasLimitBody::getGasLimit,
+                  SetGasLimit.SetGasLimitBody::setGasLimit)
+              .build();
+
+  public SetGasLimit(final Optional<BeaconProposerPreparer> beaconProposerPreparer) {
+    super(
+        EndpointMetadata.post(ROUTE)
+            .operationId("SetGasLimit")
+            .summary("Set validator gas limit")
+            .withBearerAuthSecurity()
+            .tags(TAG_GAS_LIMIT)
+            .pathParam(PARAM_PUBKEY_TYPE)
+            .description(
+                "Set the gas limit for an individual validator. This limit will be propagated to the beacon node for use on future block proposals.\n"
+                    + "The beacon node is responsible for informing external block builders of the change.\n"
+                    + "The server may return a 400 status code if no external builder is configured.\n"
+                    + "WARNING: The gas_limit is not used on Phase0 or Altair networks.")
+            .requestBodyType(GAS_LIMIT_REQUEST_BODY)
+            .response(SC_ACCEPTED, "Success")
+            .response(SC_SERVICE_UNAVAILABLE, "Unable to update gas limit at this time")
+            .withAuthenticationResponses()
+            .withNotFoundResponse()
+            .build());
+    this.beaconProposerPreparer = beaconProposerPreparer;
+  }
+
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final BLSPublicKey publicKey = request.getPathParameter(PARAM_PUBKEY_TYPE);
+    final SetGasLimit.SetGasLimitBody body = request.getRequestBody();
```

Comment:
its possible we may want some logic around bounds testing for the gas limit, but I'm not sure what that'd be ... eg. zero is likely to be not a great value to set... we could allow it I guess, but we may be setting them up for failure...

Reply:
Actually according to [the spec](https://ethereum.github.io/keymanager-APIs/#/Gas%20Limit/SetGasLimit), zero should not be accepted. The gas_limit must match the following pattern `^[1-9][0-9]{0,19}$ `
I'll add a check for that